### DISCUSSION TITLE
Bug: Records and Schemas

### DISCUSSION THREAD

**Discussion by mdocter** [microsoft/Power-Fx#388] [Q&A] [answered]
I encountered a feature in Power Apps formula's and I hope you guys can help me.

Why does the following formula

```
            If(
                true,
                {
                    onBothRecords: "First record",
                    onlyOnFirstRecord: "onlyOnFirstRecord"
                },
                {
                    onBothRecords: "Second record",
                    onlyOnSecondRecord: "onlyOnSecondRecord"
                }
            )
```
return a record with only the common field which is available on both records?
```
               {
                    onBothRecords: "First record"
               }
```
Is this behaviour documented somewhere? Where can I find more info about this behaviour?

It looks like a common schema for the records is determined, causing non-common fields to be dropped from the resulting record... 

I also asked this question in the Power Apps forum:  https://powerusers.microsoft.com/t5/Building-Power-Apps/Why-should-records-in-an-If-formula-have-the-same-fields/m-p/1576980#M404957
And MVP Sancho Harker helpt me out in the following Twitter thread: https://twitter.com/mdocter/status/1523689071574675457?t=3cVYq3mW0mEK4QJfehYvwg&s=19

**Comment by mdocter** [comment]
Update:
I did some tests and it seems that wrapping the result record in a JSON(record) formula, only shows the common properties. The Power Apps Studio also only shows the common fields in the Formula Bar preview. But when saving the result record to a data source (in my case an Oracle table) all expected fields are present...

**Comment by MikeStall** [comment, marked as answer]
Here's the pointer to the code that determines the type of an If() ... 
https://github.com/microsoft/Power-Fx/blob/28050c074b90c4df83d047bd187e81315e2098b6/src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/If.cs#L63

Along with some unit tests:
https://github.com/microsoft/Power-Fx/blob/main/src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/If.txt

If()'s resulting type is basically an intersection of the true and false branches - It doesn't know until runtime which branch will actually be, so the intersection is a safe set that will work on both the true and false branch.

**Reply by mdocter** [reply]
That's great @MikeStall. I finally understand the logic behind this behaviour. We couldn't find it in the docs and were afraid of loosing data because fields were missing. Thanks a lot!

**Reply by mdocter** [reply]
NB. After publishing and playing my test app, the JSON function still returns the intersection result... That looks like a bug.
